deckhouse-controller/pkg/helpers/change_registry/change_registry.go

```diff
@@ -326,42 +320,55 @@ func checkBearerSupport(ctx context.Context, reg name.Registry) error {
 		schemes = append(schemes, "http")
 	}
 
-	var errs []string
+	var errs *multierror.Error
 	for _, scheme := range schemes {
-		url := fmt.Sprintf("%s://%s/v2/", scheme, reg.Name())
-		req, err := http.NewRequest(http.MethodGet, url, nil)
+		u, err := url.Parse(fmt.Sprintf("%s://%s/v2/", scheme, reg.Name()))
 		if err != nil {
 			return err
 		}
-		resp, err := client.Do(req.WithContext(ctx))
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
 		if err != nil {
-			errs = append(errs, err.Error())
+			return err
+		}
+		resp, err := client.Do(req)
+		if err != nil {
+			errs = multierror.Append(errs, err)
 			// Potentially retry with http.
 			continue
 		}
-		defer func() {
-			// By draining the body, make sure to reuse the connection made by
-			// the ping for the following access to the registry
-			defer resp.Body.Close()
-			io.Copy(io.Discard, resp.Body)
-		}()
-
-		if resp.StatusCode != http.StatusUnauthorized {
-			return transport.CheckError(resp, http.StatusUnauthorized)
-		}
 
-		if challenges := authchallenge.ResponseChallenges(resp); len(challenges) != 0 {
-			// If we hit more than one, I'm not even sure what to do.
-			wac := challenges[0]
-			if strings.ToLower(wac.Scheme) == bearer {
-				return nil
-			}
-		}
-		if strings.ToLower(resp.Header.Get(wwwAuthHeader)) == bearer {
-			return nil
+		return checkResponseForBearerSupport(resp, reg.Name())
```

Comment:
```suggestion
        defer resp.Body.Close()
		return checkResponseForBearerSupport(resp, reg.Name())
```

I assume it's better to close body here, closer to the response